### PR TITLE
Processing raster calculator

### PR DESCRIPTION
**Pull request by volaya**
This PR adds a raster calculator algorithm that uses the native QGIS classes for this type of calculations. It has a custom interface and can be run from the toolbox, modeler and batch processing dialog.

When run from the console, it provides a quick alternative to using the corresponding classes. It can work on layers that are not loaded in the current project, but also with loaded layers, and in this last case it is not needed to specify the layer objects or filepaths, since they are infered from the formula itself.

Once this is merged and well-tested, I would suggest removing all other calculators from Processing (GDAL, SAGA, GRASS...), to avoid confusion. This one is versatile enough to replace the other ones, and if there is any functionality missing, it can be added to the core QgsRasterCalculator class

We should also discuss if the current raster calculator UI should be replaced with this one, like it has been done for other core algorithms (ftools, interpolation...) once an equivalent Processing algorithm has been added.

The PR includes some fixes and small changes needed to accomodate the specific characteristics of the calculator algorithm.

### CODE REVIEW

**Comment by jonnyforestGIS** [comment]
Hi @volaya my 50cent. 
I think this need to be discussed with the users. One of great thing of Processing toolbox is that we have more than one provider with the "same" algorithm, this is great because when one of the provider don't work we always have an alternative to the first one. But, by the other hand i understand your point of view because can be timming consuming keeping algorithms that do the same thing in the processing toolbox.

Regards
João

**Comment by pcav** [comment]
Thanks @volaya . One thing I believe it should be addressed is to match GRASS raster calculator, especially for proximity analyses.

**Comment by nirvn** [comment]
@jonnyforestGIS , just a week ago, I helped a QGIS user who reported not being able to use "QGIS' raster calculator". After discussing about it, it turned out she had ended up using SAGA's raster calculator (which has a issue with the raster format she was using). QGIS's native raster calculator had no problem with it. 

The problem with offering *equivalent* raster calculators (or any other algorithm) that aren't part of our source code is that a/ it doesn't integrate as well with the QGIS experience, b/ we can't react as well as we could for 3rd party algorithm bugs.

So, while it's great to offer alternatives, when it comes at the cost of negatively impacting on QGIS' experience, IMHO we need to do something about it.

**Comment by volaya** [comment]
Sometimes it's good to have alternative algorithm, in case they have something else or a different behaviour. However, in the case of this raster calculator, the idea is to have a native one that does whatever can be done with the SAGA or GRASS ones. And, since it is native, it will alwayw work and not depend on any external app. That's the reason why i am proposing to remove the other ones.

File: python/plugins/processing/algs/qgis/RasterCalculator.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        self.name, self.i18n_name = self.trAlgorithm('Raster calculator')
+        self.group, self.i18n_group = self.trAlgorithm('Raster')
+
+        self.addParameter(ParameterMultipleInput(self.LAYERS,
+                                                 self.tr('Input layers'),
+                                                 datatype=dataobjects.TYPE_RASTER,
+                                                 optional=True,
+                                                 metadata={'widget_wrapper': LayersListWidgetWrapper}))
+
+        class ParameterExpression(ParameterString):
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/RasterCalculator.py]: This should probably be renamed so it doesn't clash with https://github.com/qgis/QGIS/blob/master/python/plugins/processing/core/parameters.py#L1219 . Maybe ParameterRasterCalculatorExpression?

**Comment by nyalldawson** [comment]
@volaya 

My opinion: I'm very happy to see the qgis raster calc exposed via processing! So thanks for all the work you've done here.

> And, since it is native, it will alwayw work and not depend on any external app. That's the reason why i am proposing to remove the other ones.

I'm strongly in favor of the default calculator being QGIS' inbuilt one, for the reasons you've listed above. It's also the only one which we can directly extend, fix and optimise. IMO the core of the raster calc is solid. It needs some (complex) changes to support functions with multiple layer inputs (eg max of more than 2 bands, some form of condition if/case when support). That's the biggest thing holding it back right now. But that's all off topic anyway ;)

As far as removing the other ones... I'm not sure. I don't do enough raster calculating to feel like I have a strong enough viewpoint here or an understanding of the strengths/weakness of the other provider versions. So I'll stay out of this part of the discussion!

**Comment by pcav** [comment]
I'd suggest to check what I think is the more robust implementation:
https://grass.osgeo.org/grass73/manuals/r.mapcalc.html
Once our calc is on par with that, I think we can set it as a default one. IMHO better keep also others, even if a bit hidden from the user.

File: python/plugins/processing/algs/qgis/RasterCalculator.py (hunk 1 above)
Comment (volaya) [inline python/plugins/processing/algs/qgis/RasterCalculator.py]: I was not aware of that parameter...it seems you added it after I started working on this :-) I will change the name of my parameter, no problem

**Comment by volaya** [comment]
I think we can keep the GRASS calculator, as a more powerful alternative (GRASS is seen as an advanced thing by users anyway), and remove the SAGA and GDAL ones, which add no extra functionality.

**Comment by PedroVenancio** [comment]
Hi @volaya ,

Thanks for this effort, it's really a must-have feature for Processing, since we've this problem with r.mapcalc: http://hub.qgis.org/issues/6894

I agree with Paolo and You when say that r.mapcalc must be the reference for the new Processing raster calculator, but because of the issue #6894, it's not very helpfull keep it in Processing.

On the other hand, I don't know if you are aware of a problem/feature (?) of SAGA raster calculator, that make an interpolation when the operations involve two or more rasters, that have different resolution or different extent or different alignment. This cause, for instance, a multiplication between two integer rasters, result in a decimal raster, which was strange and difficult to understand for most users.

How will this new raster calculator tool deal with that question (input rasters with different resolution, different extent and/or different alignment)?

Finally I agree with you, after this tool is well tested, the other ones can be removed, since none of them work perfectly at this time.

Thank you very much!

**Comment by alexbruy** [comment]
Travis passed for PR, failure was for the branch in the QGIS repo created by mistake. So I merged it.

**Comment by nirvn** [comment]
@volaya , @alexbruy , nice!

There's a bit of an issue with the UI, see screenshot below:
untitled
* "Layers (double-click...)" and "Expression" labels overlap with other widgets
* The values and operators buttons' width could be reduced a bit to avoid making the whole dialog overly large.

**Comment by pcav** [comment]
Confirmed, resizing windows gives suboptimal results.
Wonderful work, thanks!

**Pull request merged**